core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -743,28 +742,23 @@ private[zio] final class FiberContext[E, A](
     val oldState = state.get
 
     oldState match {
-      case Executing(FiberStatus.Suspended, observers0) if interruptible =>
+      case Executing(FiberStatus.Suspended(true), observers0, false) =>
         val observers = k :: observers0
 
-        if (!state.compareAndSet(oldState, Executing(FiberStatus.Running, observers))) kill0(k)
+        if (!state.compareAndSet(oldState, Executing(FiberStatus.Running, observers, true))) kill0(k)
         else {
-          interrupted = true
-
           evaluateLater(ZIO.interrupt)
-
           None
         }
 
-      case Executing(status, observers0) =>
+      case Executing(status, observers0, _) =>
```

Comment:
This extra flag is necessary in your fix. In my fix the model changed a bit, so pushing the "set" prior to the atomic ref update should in theory take care of it. But this looks good for now!